Mage/src/main/java/mage/abilities/keyword/GiftAbility.java

```diff
@@ -0,0 +1,174 @@
+
+/**
+ * @author TheElk801
+ */
+public class GiftAbility extends StaticAbility implements OptionalAdditionalSourceCosts {
+
+    private static final String keywordText = "Gift";
+    private static final String reminderText = "You may promise an opponent a gift as you cast this spell. If you do, ";
+    private final String rule;
+
+    public static final String GIFT_ACTIVATION_VALUE_KEY = "giftPromisedActivation";
+
+    protected OptionalAdditionalCost additionalCost;
+    private final GiftType giftType;
+
+    private static String makeReminderText(GiftType giftType, boolean isPermanent) {
+        return reminderText + (isPermanent ? "when it enters, " : "") + "they " +
+                giftType.getDescription() + (isPermanent ? "." : " before its other effects.");
+    }
+
+    public GiftAbility(Card card, GiftType giftType) {
+        super(Zone.STACK, null);
+        this.additionalCost = new OptionalAdditionalCostImpl(
+                keywordText + ' ' + giftType,
+                makeReminderText(giftType, card.isPermanent()),
+                new PromiseGiftCost(giftType)
+        );
+        this.additionalCost.setRepeatable(false);
+        this.giftType = giftType;
+        this.rule = additionalCost.getName() + ' ' + additionalCost.getReminderText();
+        this.setRuleAtTheTop(true);
+        if (card.isPermanent()) {
+            this.addSubAbility(new ConditionalInterveningIfTriggeredAbility(
+                    new EntersBattlefieldTriggeredAbility(new PromiseGiftEffect(giftType)),
+                    GiftWasPromisedCondition.instance, "When this permanent enters, " +
+                    "if the gift was promised, they " + giftType.getDescription() + '.'
+            ).setRuleVisible(false));
+        } else {
+            card.getSpellAbility().addEffect(new PromiseGiftEffect(giftType));
+        }
+    }
+
+    private GiftAbility(final GiftAbility ability) {
+        super(ability);
+        this.rule = ability.rule;
+        this.additionalCost = ability.additionalCost.copy();
+        this.giftType = ability.giftType;
+    }
+
+    @Override
+    public GiftAbility copy() {
+        return new GiftAbility(this);
+    }
+
+    @Override
+    public void addOptionalAdditionalCosts(Ability ability, Game game) {
+        if (!(ability instanceof SpellAbility)) {
+            return;
+        }
+        Player player = game.getPlayer(ability.getControllerId());
+        if (player == null) {
+            return;
+        }
+        additionalCost.reset();
+        if (!additionalCost.canPay(ability, this, ability.getControllerId(), game)
+                || !player.chooseUse(Outcome.Benefit, "Promise an opponent " + giftType.getName() + '?', ability, game)) {
+            return;
+        }
+        additionalCost.activate();
+        for (Cost cost : ((Costs<Cost>) additionalCost)) {
+            ability.getCosts().add(cost.copy());
+        }
+        ability.setCostsTag(GIFT_ACTIVATION_VALUE_KEY, null);
```

Comment:
hm, so this gets set both here and in `pay()` ? Is it because you need to choose alt costs, then select targets which may depend on alt costs, then pay those costs? If so, let's add some comments explaining.

Reply:
@ssk97 the tests still pass without this line, should I just remove it?

Reply:
I think that the cards where the target changes based on the gift being given or not will fail if this line is removed, since the tag won't be set when the target adjuster is called.

Reply:
oh that's right, good point

Reply:
Let's get some test coverage on that then